feat: [M3-7114] - Allow user to choose resize migration type when resizing Linode

jaalah-akamai opened the pull request:
## Description 📝
This work will introduce `warm` migrations which combine the benefits of `cold` and `live` migrations to greatly reduce customer downtime when resizing Linodes.

> **Important**
The notification and event copy/messaging for `warm` migrations will be updated with API work, but you test the UI flow, payload, feature flags and everything else. I will have to update the cypress tests later with actual copy (9/27)

## Major Changes 🔄

- [x] Add feature flag `unifiedMigrations`

**Resize Flow:**
- [x] Add an optional `migration_type` POST field for the `linode/instances/.../resize` endpoint that should be `warm` or `cold`. Backend defaults to `cold`, but in the UI we are defaulting to `warm` unless the user chooses otherwise.
- [x] Add a radio button group for selecting `warm` or `cold` migrations in the `LinodeResize` flow.
    - [x] If the Linode is powered off, we disable controls (even though warm is selected by default) and proceed with a `cold` migration.
- [x] Update & add cypress E2E tests for changes

**Linode Landing:**
- [x] As a user the flag icon is ambiguous and doesn't give much meaning, therefore we want to remove the icon altogether and provide and updated status where it makes sense. For resizes, we want to change the wording from `running` to `migrating` with a green dot icon


**Other:**
- [x]  Fixed the animations for `<Notice />` and fixed a bug in our `fadeIn` keyframes. _Pesky colons!_
    - [x]  In addition, the padding for the `<Notice />` didn't seem right or match `<DismissibleBanner />` which seemed to be the intention, so this has been fixed.

## TODO:
- [ ] Waiting on APIv4 dev work with event/notification messaging to update Cypress tests
- [x] Waiting on UX copy for messaging

## Preview 📷
| Before  | After   |
| ------- | ------- |
| ![Screenshot 2023-09-13 at 9 30 01 PM](https://github.com/linode/manager/assets/125309814/f9c0bba4-6230-41b7-a649-8221c5342d39) | ![Screenshot 2023-09-13 at 9 29 52 PM](https://github.com/linode/manager/assets/125309814/ea9e5c2d-0da4-4522-8d2d-85fd05cd3e02) |

|Event Notification|
| ------- | 
|<img width="839" alt="Screenshot 2023-09-15 at 11 56 01 AM" src="https://github.com/linode/manager/assets/125309814/962d93fd-207c-4747-b0ee-05b8a489f751">|

| Enabled  | Disabled   |
| ------- | ------- |
| ![Screenshot 2023-09-28 at 9 29 14 AM](https://github.com/linode/manager/assets/125309814/6efa1bba-9684-447d-b052-8dc6945b4189) | ![Screenshot 2023-09-28 at 9 27 39 AM](https://github.com/linode/manager/assets/125309814/6b97fb88-ba18-47ae-a05c-12e10b9a248c) |

| Payload  |
| ------- |
|![Screenshot 2023-09-13 at 3 49 32 PM](https://github.com/linode/manager/assets/125309814/c6f4eb67-c158-4f66-b898-d12bc8f17fb3)|

| Linode Landing  |
| ------- |
|![Screenshot 2023-09-27 at 10 34 27 PM](https://github.com/linode/manager/assets/125309814/d7661f27-6062-4376-be4a-5806ca849283) ![Screenshot 2023-09-27 at 10 30 26 PM](https://github.com/linode/manager/assets/125309814/83326319-2a8a-4aa1-94ab-18b0917051db)|

| Feature Flag  |
| ------- |
|![Screenshot 2023-09-25 at 7 45 22 PM](https://github.com/linode/manager/assets/125309814/055af4f4-5540-49ee-a2be-d1053783b82c)|


## How to test 🧪
1. Enable `Unified Migrations` feature flag via dev tools
2. Go to Linodes landing page and resize a Linode from the action menu
3. Observe you now have a radio group to choose `warm` or `cold` migrations with `warm` set as the default
4. When pressing `Resize Linode` you should see a confirmation dialog appear before resize takes effect
5. During resize we should see resize events as outlined in screenshot and notification (notice) on Linode detail page.
    1. Note: The copy for a `warm` resize has not been updating yet (awaiting API), but you should see the payload with the new `migration_type: warm` as shown in screenshot.
6. A `cold` resize does NOT have any additional confirmation
7. Confirm the confirmation dialog closes on error + the resize panel scrolls to the error for a warm migration; confirm that the resize panel scrolls to the error for a cold migration
8. Confirm that errors do not persist upon closing/reopening the resize panel (for both the same linode and a different linode)
9. Confirm that if a linode is offline, default is cold migration + the panel is disabled and a tooltip appears. Confirm that even if you close/reopen the resize panel, the default stays as the cold migration
10. Confirm the feature flag works as expected
11. Confirm the 'Migration Types' label in the migrations panel is not highlighted when choosing between the migration types

mjac0bs commented:
> In addition, the padding for the <Notice /> didn't seem right or match <DismissibleBanner /> which seemed to be the intention, so this has been fixed.

Thanks for fixing this! 🎉 🥇  I'd recently noticed that a few notices were looking weirdly skinny, seemingly randomly, but hadn't tracked it down or made a ticket.

jdamore-linode commented on packages/manager/cypress/e2e/core/linodes/resize-linode.spec.ts:
```diff
@@ -16,9 +26,88 @@ describe('resize linode', () => {
+    });
+  });
+
+  it.only('resizes a linode by increasing size: cold migration', () => {
```
```suggestion
  it('resizes a linode by increasing size: cold migration', () => {
```

jaalah-akamai replied:
Thanks! Forgot to revert this from testing!

jdamore-linode reviewed (commented):
Hoping to circle back soon with a proper review, but just swinging in now to thank you for the Cypress test additions!

coliu-akamai commented on packages/api-v4/src/linodes/actions.ts:
```diff
@@ -77,10 +77,13 @@ export const linodeShutdown = (linodeId: number | string) =>
+ * @param data.allow_auto_disk_resize { boolean } do you want to resize your disks after
  * the Linode is resized? NOTE: Unless the user has 1 ext disk or 1 ext disk and
- * 1 swap disk, this flag does nothing, regardless of whether it's true or false
+ * 1 swap disk, this flag 4does nothing, regardless of whether it's true or false
```
```suggestion
 * 1 swap disk, this flag does nothing, regardless of whether it's true or false
```

coliu-akamai reviewed (commented):
✅ Observe you now have a radio group to choose warm or cold migrations 
✅ Warm is set as default
✅  When pressing Resize Linode you should see a confirmation dialog appear before resize takes effect 
✅ During resize we should see resize events as outlined in screenshot and notification (notice) on Linode detail page. Note: The copy for a warm resize has not been updating yet (awaiting API), but you should see the payload with the new migration_type: warm as shown in screenshot. 
✅ A cold resize does NOT have any additional confirmation (assuming this means no confirmation dialog)

Something interesting to note: when I clicked 'Continue' in the confirmation popup, if an error appears, the popup stays -- not sure if this is the behavior you want or not:
![image](https://github.com/linode/manager/assets/139280159/3fd67769-e20c-477b-b7bf-709224fef07f)

abailly-akamai reviewed (changes requested):
Nice! looks great overall.

A couple comments on the UI:

This tooltip should be wider. The copy is a bit lengthy and It's pretty hard to read currently.
![Screenshot 2023-09-27 at 17 30 32](https://github.com/linode/manager/assets/130582365/79c4cbfc-e649-4e34-9d63-a25513ddafb8)

The logic for disabling the field if a linode is powered down is implemented, however there's no explanation as to why. Yes, it's logical we can't perform a warm migration on a powered down Linode, but I don't think we should assume a user will immediately figure out (it's a new feature). I think the UX is lacking here and I'd rather see the "Cold" selection enabled with no option to switch to the warm option and some helper text as to why.

jaalah-akamai commented:
@abailly-akamai:
- I finished updating our `TooltipIcon` to be more flexible and allow a `width` to be passed in for variable widths, and then I realized the latest mock excludes a tooltip altogether. Anyways I kept it in because I think it could be useful and added a storybook example. 
- I agree with your second concern and I sent it over to UX. We either need some explanation or we just set it to `cold` in the disabled state. I'm awaiting their reply! Hopefully it doesn't hold up this PR because it will need to be in next release. 🤞 

![Screenshot 2023-09-27 at 9 20 21 PM](https://github.com/linode/manager/assets/125309814/758af3d2-a424-40a0-8286-c00fad6a143b)

@coliu-akamai The dialog should close on error and scroll to the top still, good catch 👍

jaalah-akamai commented:
@abailly-akamai I also want to point out some additional changes for the Linodes Landing page (updated description):

- [47cc795](https://github.com/linode/manager/pull/9677/commits/47cc795ffac9f0db275504267e749e9d7a3ceac6)
- [72407cf](https://github.com/linode/manager/pull/9677/commits/72407cfb42fa5bf5b1ae66a21ea22bcbf0c991b8)

coliu-akamai commented:
Will be checking these off as I go through them
- [ ] Observe you now have a radio group to choose warm or cold migrations
- [ ] Warm is set as default
- [ ] When pressing Resize Linode you should see a confirmation dialog appear before resize takes effect
- [ ] During resize we should see resize events as outlined in screenshot and notification (notice) on Linode detail page. Note: The copy for a warm resize has not been updating yet (awaiting API), but you should see the payload with the new migration_type: warm as shown in screenshot.
- [ ] A cold resize does NOT have any additional confirmation (assuming this means no confirmation dialog)
- [ ] dialog closes on error
- [ ] if linode is offline, default is cold migration + panel disabled

coliu-akamai commented:
Are errors supposed to be cleared if we close the panel and then reopen? The following error is persistent (which honestly I could see making sense for at least this specific error but maybe not other errors -- just wanted to double check):
![image](https://github.com/linode/manager/assets/139280159/20959705-8779-4946-af5d-1d4c655aa546)

abailly-akamai commented:
@jaalah thx for the changes - looks much better 👍 

I also noticed there's an issue with persisting state in the resize modal which I think is unrelated to this PR, but which this PR makes more evident. See video below, look at the persisting error banner and new modal across various resizes attempts

https://github.com/linode/manager/assets/130582365/b7ba9001-db01-466d-b14a-810a76edb1f9

dwiley-akamai commented on packages/api-v4/.changeset/pr-9677-added-1695693363239.md:
```diff
@@ -0,0 +1,5 @@
+"@linode/api-v4": Added
+---
+
+New event type `linode_resize_warm_create` and payload option `migration_type` ([#9677](https://github.com/linode/manager/pull/9677))
```
```suggestion
 New payload option `migration_type` in `ResizeLinodePayload` and new event type `linode_resize_warm_create`([#9677](https://github.com/linode/manager/pull/9677))
```

abailly-akamai replied:
done

dwiley-akamai commented on packages/manager/src/features/Linodes/LinodesDetail/LinodeResize/LinodeResize.tsx:
```diff
@@ -189,6 +251,74 @@ export const LinodeResize = (props: Props) => {
+              }}
+              component="p"
+            >
+              During a <strong>warm resize</strong>, your Compute Instance will
```
We're inconsistent in the casing for "warm resize" in the UI copy -- should it be standardized as either "Warm resize" or "warm resize" across the board?

coliu-akamai replied:
ux confirmed -- should be lowercase across the board (except for the radio button labels - those will say 'Warm resize' or 'Cold resize'), so made those changes!

dwiley-akamai reviewed (commented):
- Feature flag functionality ✅
- Radio button group for selecting warm or cold migration in the Linode Resize modal ✅
  - Radio button group disabled if Linode is powered off ✅
  - Confirmation modal displayed if “Warm resize” chosen ✅
- Events/Notifications appear as expected ✅

When you click one of the radio buttons, the label gets highlighted:

https://github.com/linode/manager/assets/114682940/eb88001b-5718-4746-ac04-45c30e50a6ae

I'm also observing linodes not having their statuses updated properly after resizes in the Linodes landing table (e.g., resize one that was running --> the process completes and the linode shows as "Offline" in the table, although it's actually running; it doesn't update properly until a refresh), but that seems to be an existing issue?

coliu-akamai reviewed (commented):
Confirmed that the issue with the error persisting is gone!
Holding off on approval while we're investigating this weird issue some more:

![image](https://github.com/linode/manager/assets/139280159/2584dd06-5d75-4d62-aa54-1d377246f3af)

abailly-akamai reviewed (approved):
Leaving my approval here tho it shouldn't count as one of the co-author.

Tested all scenari i could think of with a special focus on the dialog state management.

coliu-akamai reviewed (approved):
(Same boat with Alban, this approval probably shouldn't count)

✅ Observe you now have a radio group to choose `warm` or `cold` migrations with `warm` set as the default
✅ When pressing `Resize Linode` you should see a confirmation dialog appear before resize takes effect
✅ During resize we should see resize events as outlined in screenshot and notification (notice) on Linode detail page. Note: The copy for a `warm` resize has not been updating yet (awaiting API), but you should see the payload with the new `migration_type: warm` as shown in screenshot.
✅ A `cold` resize does NOT have any additional confirmation
✅  Confirm the confirmation dialog closes on error + scrolls
✅  Confirm that errors do not persist upon closing/reopening the resize panel (for both the same linode and a different linode)
✅  Confirm that if a linode is offline, default is cold migration + panel disabled. Confirm that even if you close/reopen the resize panel, the default stays as the cold migration
✅  Confirm the feature flag works as expected
✅  Confirm the 'Migration Types' label in the migrations panel is not highlighted when choosing between the migration types

Creating a separate ticket for styling of disabled radio buttons in dark theme
![image](https://github.com/linode/manager/assets/139280159/9da0ccb7-35ac-4686-aad7-2e028311103c)

bnussman-akamai reviewed (approved):
Tested general functionality and everything worked!

I did notice a few things (non-blocking 🚧)
- I don't like that the Migration type is in its own Paper
- The Linode row shows the text "Migrating" instead of "Resizing" during resize
![Screenshot 2023-09-28 at 4 14 28 PM](https://github.com/linode/manager/assets/115251059/0f9a8228-fc58-4b88-bbfe-c81180297e6d)


- The general error banner is extra thick
![Screenshot 2023-09-28 at 4 10 31 PM](https://github.com/linode/manager/assets/115251059/568dcb24-6dfb-404d-8a6e-b954917549dc)

dwiley-akamai reviewed (approved):
Errors don't persist between dialog openings/closings ✅
UI bug of radio group label fixed ✅

Going ahead and approving this but we may need a follow-up ticket/PR for the discrepancies below:

> [x] As a user the flag icon is ambiguous and doesn't give much meaning, therefore we want to remove the icon altogether and provide and updated status where it makes sense. For resizes, we want to change the wording from `running` to `migrating` with a green dot icon

Just to confirm, was this supposed to be "Resizing" --> "Migrating"? In prod we have "Resizing" with a blinking yellow dot. 

Also, should the dot on the Linode Detail page be updated to match what's in the Linodes landing table when the migration is happening? Right now we have "Migrating (xy%)" in the Linodes landing table but the Linode Detail page displays "Resizing":

<details>
<summary>Screenshots</summary>

![Screenshot 2023-09-28 at 4 16 41 PM](https://github.com/linode/manager/assets/114682940/85505f8d-993b-4a5a-bca4-0aaef9c6e28d)

![Screenshot 2023-09-28 at 4 16 51 PM](https://github.com/linode/manager/assets/114682940/8e3c406c-585e-4474-a996-3d04ebdb1a6d)
</details>

abailly-akamai commented:
@dwiley-akamai & @bnussman-akamai I am thinking these items should be addressed. Taking a look and asking for feedback.

My understanding is that we're indeed calling this a migration now, but I agree it does not make sense to call it that while everything else is calling it a resize. Consistency should be key

cpathipa commented:
IMO, it should be consistent, Also in linodes details page the font style is all caps ? it should be "Resizing"
![image](https://github.com/linode/manager/assets/119517080/918afb9c-09e1-4e18-91b6-8387e5930b7d)

abailly-akamai commented:
@mjac0bs @jaalah-akamai 

just to confirm - do we really want to make ALL our notices that fat?

![Screenshot 2023-09-29 at 11 34 38](https://github.com/linode/manager/assets/130582365/729c98c5-e822-432f-ab81-f089ee43fb41)

CC @bnussman-akamai https://github.com/linode/manager/pull/9677#pullrequestreview-1649641530

abailly-akamai commented:
Notice size change has been reverted in this PR and can be revisited later. We're ready to merge as need be.

cpathipa commented:
@abailly-akamai QQ: During a cold resize, it mentions that the instance will be shut down. In that case, we expected to see a green status icon or an orange one.? 
![image](https://github.com/linode/manager/assets/119517080/51f3103e-8243-46b1-b953-3cf12e3eedb5)

abailly-akamai commented:
@cpathipa it does for me

![Screenshot 2023-09-29 at 13 07 27](https://github.com/linode/manager/assets/130582365/137ae6e0-a828-4067-b9c6-79982d93eec7)

cpathipa commented:
@abailly-akamai It looks like the issue is inconsistent. You can notice the green status at 2:04 in the clip. Looking into that..

https://github.com/linode/manager/assets/119517080/ca83e6fc-d19e-4831-ad13-58a149be62ed

bnussman-akamai commented:
@cpathipa This may be happening because we don't invalidate Linodes when an event is `started`. I don't know a way off the top of my head to solve this without overfetching

https://github.com/linode/manager/blob/develop/packages/manager/src/queries/linodes/events.ts#L23-L25

bnussman-akamai commented on packages/api-v4/src/linodes/types.ts:
```diff
@@ -394,10 +394,14 @@ export interface LinodeDiskCreationData {
   type: string;
   /** @default true */
   allow_auto_disk_resize?: boolean;
+  /** @default 'warm' */
+  migration_type?: MigrationTypes;
```
Which is the actual correct default?

![Screenshot 2023-10-02 at 1 15 05 PM](https://github.com/linode/manager/assets/115251059/80acfbac-ec9d-40ad-bd47-c4e06563cc2c)

Based on this comment in the PR description, it should be `cold`

abailly-akamai replied:
correct, only the ui defaults to warm. will update

jdamore-linode reviewed (approved):
Pushed a quick fix for the one failing resize test! Looking good to me 👍

jaalah-akamai commented:
![Screenshot 2023-10-10 at 9 29 04 AM](https://github.com/linode/manager/assets/125309814/eb7b6f95-cdb2-4bfa-8b69-75106373b97d)

- Moved `(Recommended)` outside of tooltip next to label text
- Added `Learn more.` link to documentation for both cold/warm
